### DISCUSSION TITLE
Bump PHPStan analysis level

### DISCUSSION THREAD

**Discussion by Wirone** [PHP-CS-Fixer/PHP-CS-Fixer#6867] [RFC]
Currently we have PHPStan analysis on level 6, which is relatively low (max is 9). I believe that rising level can help with producing better code and catch more flaws. **Current code on level 9 causes ~3k errors to be reported**.

As a reference: in GetResponse we currently [bumped level to 9](https://twitter.com/_Codito_/status/1638489936893550593) which caused 36k+ of errors to be added to baseline files. Root cause of that decision is shown [here](https://phpstan.org/r/5efc642e-920b-4598-ab68-b7be223e3011) - because one error was ignored on level 6 it could lead to other problems that were reported only on level 9...

Bumping level and dumping existing errors to baseline can allow us to catch more potential issues in newly created code, while keeping existing issues as backlog to fix. We could enforce that baseline can only shrink somehow and clean it from time to time.

**Comment by julienfalque** [comment]
If I remember correctly we already tried that, but it was a bit painful to maintain the baseline. I'm OK to try again anyway.

> We could enforce that baseline can only shrink

Sometimes, fixing an error allows PHPStan to better understand a part of the code and start to detect more errors in that code, hence making the baseline grow. In this case this should be allowed, otherwise one would have to fix all newly detected errors, even if they were not actually introduced with the changes.

**Reply by Wirone** [reply]
Yes, there are scenarios where new rules in baseline are allowed, it always depends on the context 😉.

FYI: baseline approach is going to be changed [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6868) (introducing proper baseline file instead of defining ignore rules in config file + reporting of unmatched errors was enabled). Maintaining it requires only 1 command for dumping current set of errors (can be also defined as Composer script [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6839).
